kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/sortedDesc.kt

```diff
@@ -26,33 +61,76 @@ infix fun <T, E : Comparable<E>> List<T>.shouldBeSortedDescendingBy(transform: (
 }
 
 infix fun <T, I : Iterable<T>, E : Comparable<E>> I.shouldBeSortedDescendingBy(transform: (T) -> E): I {
-   toList().shouldBeSortedDescendingBy(transform)
+   this should beSortedDescendingBy(transform)
    return this
 }
 
-fun <T : Comparable<T>> beSortedDescending(): Matcher<List<T>> = sortedDescending()
+fun <T : Comparable<T>> beSortedDescending(): Matcher<Iterable<T>> = beSortedDescending(null)
+
+fun <T : Comparable<T>> sortedDescending(): Matcher<Iterable<T>> = beSortedDescending()
+
+fun <T, E : Comparable<E>> beSortedDescendingBy(transform: (T) -> E): Matcher<Iterable<T>> =
+   beSortedDescendingWith(null) { a, b -> transform(a).compareTo(transform(b)) }
+
+fun <T, E : Comparable<E>> sortedDescendingBy(transform: (T) -> E): Matcher<Iterable<T>> =
+   beSortedDescendingBy(transform)
 
-fun <T : Comparable<T>> sortedDescending(): Matcher<List<T>> = sortedDescendingBy { it }
+private fun <T : Comparable<T>> beSortedDescending(name: String?): Matcher<Iterable<T>> =
+   beSortedDescendingWith(name) { a, b -> a.compareTo(b) }
 
-fun <T, E : Comparable<E>> beSortedDescendingBy(transform: (T) -> E): Matcher<List<T>> = sortedDescendingBy(transform)
+private fun <T, I : Iterable<T>> beSortedDescendingWith(name: String?, comparator: Comparator<T>): Matcher<I> =
+   object : Matcher<I> {
+      override fun test(value: I): MatcherResult {
+         val name = name ?: value.containerName()
+         val positiveResult = MatcherResult(
+            true,
+            { "$name should be sorted descending" },
+            { "$name should not be sorted descending" }
+         )
 
-fun <T, E : Comparable<E>> sortedDescendingBy(transform: (T) -> E): Matcher<List<T>> = object : Matcher<List<T>> {
-   override fun test(value: List<T>): MatcherResult {
-      val failure =
-         value.withIndex().firstOrNull { (i, it) -> i != value.lastIndex && transform(it) < transform(value[i + 1]) }
-      val elementMessage = when (failure) {
-         null -> ""
-         else -> ". Element ${failure.value} at index ${failure.index} was less than element ${value[failure.index + 1]}"
+         value.zippedWithNext(
+            onEmpty = { return positiveResult },
+            onSingle = { return positiveResult }
+         ) { a, b ->
+            if (comparator.compare(a.value, b.value) < 0) {
+               return MatcherResult(
+                  false,
+                  { "$name should be sorted in descending order. Element ${a.value.print().value} at index ${a.index} was smaller than element ${b.value.print().value} at index ${b.index}" },
+                  { "$name should not be sorted descending" }
+               )
+            }
+         }
+
+         return positiveResult
       }
-      return MatcherResult(
-         failure == null,
-         { "List ${value.print().value} should be sorted$elementMessage" },
-         { "List ${value.print().value} should not be sorted" }
-      )
+   }
+
+internal inline fun <T> Iterable<T>.zippedWithNext(
```

Comment:
Is there a reason the following short idiomatic one-liner was expanded into a 20-line  non-idiomatic method:

`". Element ${failure.value} at index ${failure.index} was less than element ${value[failure.index + 1]}"`

Let's revert back to that simple one-liner.

Reply:
Iterable isn't a `RandomAccess` container, so it has neither the `get` operator nor `size ` or `lastIndex`.

This algorithm performs a single pass over the container without extraneous allocations and will be reused for both ascending `beSorted` matchers and sequence `beSorted` matchers.

Reply:
> Iterable isn't a `RandomAccess` container, so it has neither the `get` operator nor `size ` or `lastIndex`.
> 
> This algorithm performs a single pass over the container without extraneous allocations and will be reused for both ascending `beSorted` matchers and sequence `beSorted` matchers.

Kotlin's own `firstOrNull` does all that as well, so please use it instead or reimplementing.

Reply:
> > Iterable isn't a `RandomAccess` container, so it has neither the `get` operator nor `size ` or `lastIndex`.
> > This algorithm performs a single pass over the container without extraneous allocations and will be reused for both ascending `beSorted` matchers and sequence `beSorted` matchers.
> 
> Kotlin's own `zipWithNext` and `firstOrNull` do all that as well, so please use them instead or reimplementing.

No, it doesn't - `zipWithNext` is a terminal operation which buffers the entire Iterable in memory by allocating a new list of the resulting pairs. If it did do the same as the introduced function, I would have used it. Do you want to forgo the single-pass algorithm in favor of the extraneous allocation?

Reply:
you can use this one:

https://github.com/JetBrains/kotlin/blob/eb6e6b7b526f06ff72b56de3c7949d91f84d50e6/libraries/stdlib/common/src/generated/_Sequences.kt#L2851